File: sdks/python/aegis_sdk/resources/reports.py
"""Reports resource for the Aegis SDK."""

from __future__ import annotations

from typing import Any

import httpx

from aegis_sdk.models import (
    CategoryBreakdown,
    MonthlyTrend,
    NetWorth,
    YearlySummary,
)


class ReportsResource:
    """Provides access to report endpoints (``/api/reports``)."""

    _path = "/api/reports"

    def __init__(self, client: httpx.Client) -> None:
        self._client = client

    def monthly_trend(self, *, months: int | None = None) -> list[MonthlyTrend]:
        """Get monthly income/expense trends.

        Args:
            months: Number of months to include in the trend.
        """
        params: dict[str, Any] = {}
        if months is not None:
            params["months"] = months
        resp = self._client.get(f"{self._path}/monthly-trend", params=params)
        resp.raise_for_status()
        return [MonthlyTrend.model_validate(item) for item in resp.json()]

    def category_breakdown(
        self, *, month: str | None = None
    ) -> list[CategoryBreakdown]:
        """Get spending breakdown by category.

        Args:
            month: Month to filter by (e.g. ``"2026-03"``).
        """
        params: dict[str, Any] = {}
        if month is not None:
            params["month"] = month
        resp = self._client.get(f"{self._path}/category-breakdown", params=params)
        resp.raise_for_status()
        return [CategoryBreakdown.model_validate(item) for item in resp.json()]

    def yearly_summary(self, *, year: int | None = None) -> YearlySummary:
        """Get a yearly financial summary.

        Args:
            year: Calendar year (defaults to current year on the server).
        """
        params: dict[str, Any] = {}
        if year is not None:
            params["year"] = year
        resp = self._client.get(f"{self._path}/yearly-summary", params=params)
        resp.raise_for_status()
        return YearlySummary.model_validate(resp.json())

    def net_worth(self) -> NetWorth:
        """Get the current net-worth calculation."""
        resp = self._client.get(f"{self._path}/net-worth")
        resp.raise_for_status()
        return NetWorth.model_validate(resp.json())

File: sdks/typescript/src/resources/reports.ts
import type { AegisClient } from "../client";
import type {
  MonthlyTrendEntry,
  CategoryBreakdown,
  YearlySummary,
  NetWorth,
} from "../types";

export class ReportsResource {
  private client: AegisClient;

  constructor(client: AegisClient) {
    this.client = client;
  }

  /** Get monthly income/expense trends. */
  async getMonthlyTrend(params?: { months?: number }): Promise<MonthlyTrendEntry[]> {
    return this.client.get<MonthlyTrendEntry[]>("/api/reports/monthly-trend", {
      months: params?.months,
    });
  }

  /** Get spending breakdown by category. */
  async getCategoryBreakdown(params?: { month?: string }): Promise<CategoryBreakdown> {
    return this.client.get<CategoryBreakdown>("/api/reports/category-breakdown", {
      month: params?.month,
    });
  }

  /** Get a yearly financial summary. */
  async getYearlySummary(params?: { year?: number }): Promise<YearlySummary> {
    return this.client.get<YearlySummary>("/api/reports/yearly-summary", {
      year: params?.year,
    });
  }

  /** Get the current net-worth calculation. */
  async getNetWorth(): Promise<NetWorth> {
    return this.client.get<NetWorth>("/api/reports/net-worth");
  }
}
